File: ArtificialIntelligence/hw1_notworking/GameModels/PegSolitaireState.py
#This class is a container of the state of the game. It contains various, general information about it

import GameModels as G
class PegSolitaireState:
    
    lastID = 1 #needed for debug

    def __init__(self, heuristic, grid, parentState = None):
        self.H = heuristic
        self.representation = G.PegSolitaireRepresentation(grid)
        self.parentState = parentState
        
        PegSolitaireState.lastID += 1
        self.identifier = PegSolitaireState.lastID

    def getRepresentation(self):
        return self.representation

    def getHeuristic(self):
        return self.H
    
    def parent(self):
        return self.parentState
    
    #needed for debug
    def getStateName(self):
        stateName = str(self.level) + ", id= " + str(self.identifier)
        return stateName
        
    def setLevel(self, level):
        self.level = level
        
    def getLevel(self):
        return self.level
    
    def __eq__(self, other):
        return self.representation == other.representation
    
    #needed for set (add and remove) and list (remove)
    def __hash__(self):
        grid = self.representation.grid
        l = ""
        for j in range(0,len(grid)):
            for k in range(0,len(grid[0])):
                if grid[j][k] != 2:
                    l += str(grid[j][k])
        h = int(l)
        
#        h = hash(self.representation)
        return h
       
    def __repr__(self):
        return repr(self.representation)   
        

File: ArtificialIntelligence/hw1java/src/GameModel/PegSolitaireState.java
package GameModel;

import Heuristic.Heuristic;

public class PegSolitaireState implements State {

	private static int lastID = 1; //needed for debug
	private Heuristic H;
	private Representation representation;
	private PegSolitaireState parentState;
	private int identifier;
	private int level;

    public PegSolitaireState(Heuristic heuristic, int[][] grid, PegSolitaireState parentState ) {
        this.H = heuristic;
        this.representation = new PegSolitaireRepresentation(heuristic, grid);
        this.parentState = parentState;
        
        PegSolitaireState.lastID += 1;
        this.identifier = PegSolitaireState.lastID;
    }

    public Representation getRepresentation() {
        return this.representation;
    }

    public Heuristic getHeuristic() {
        return this.H;
    }
    
    public PegSolitaireState parent() {
        return this.parentState;
    }
    
    //needed for debug
    public String getStateName() {
        String stateName = "level= " + this.level + ", id= " + this.identifier;
        return stateName;
    }
        
    public void setLevel(int level) {
        this.level = level;
    }
        
    public int getLevel() {
        return this.level;
    }
    
    //needed for hashset
    @Override
    public boolean equals(Object other) {
    	PegSolitaireState otherstate = (PegSolitaireState) other;
        return (this.representation).equals(otherstate.representation);
    }
    
    @Override
    public int hashCode() {
    	int h =  ((PegSolitaireRepresentation) this.representation).hashCode();
    	return h;
    }

	@Override
	public int f() {
        int g = 1; //could be any number. this is the cost to move a peg
        int h = this.H.H((State) this);
        int valf = g + h;
        return valf;
    }

	public boolean isSolution() {
		return ((PegSolitaireRepresentation) this.getRepresentation()).isSolution();
	}
	
	public String toString() {
		int[][] grid = ((PegSolitaireRepresentation) this.representation).grid();
		String r = "";
		for (int i = 0; i < grid.length; i++) {
            for ( int j = 0; j < grid[0].length; j++)
            	r += grid[i][j] + " ";
            r += "\n";
		}
		return r;
	}

	public void setParent(PegSolitaireState parent) {
		this.parentState = parent;
	}
                
}
